pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -1090,12 +1090,29 @@ func (m *ManagerImpl) isDevicePluginResource(resource string) bool {
 	return false
 }
 
-// GetAllocatableDevices returns information about all the devices known to the manager
 func (m *ManagerImpl) GetAllocatableDevices() ResourceDeviceInstances {
 	m.mutex.Lock()
-	resp := m.allDevices.Clone()
-	m.mutex.Unlock()
-	klog.V(4).InfoS("Known devices", "numDevices", len(resp))
+	defer m.mutex.Unlock()
+	klog.V(4).InfoS("Known devices", "numDevices", len(m.allDevices))
+
+	resp := NewResourceDeviceInstances()
+	for resourceName, deviceInstances := range m.allDevices {
+		healthyIDs, ok := m.healthyDevices[resourceName]
```

Comment:
I am just curious, how does the device manager decide when the device is healthy and unhealthy?

Reply:
This callback https://github.com/kubernetes/kubernetes/blob/a2e12354edefc420e829973721030c5cb2a4270d/pkg/kubelet/cm/devicemanager/manager.go#L180 takes care of updating the health status. It is registered in the endpoint (see: https://github.com/kubernetes/kubernetes/blob/a2e12354edefc420e829973721030c5cb2a4270d/pkg/kubelet/cm/devicemanager/manager.go#L165 >>> https://github.com/kubernetes/kubernetes/blob/a2e12354edefc420e829973721030c5cb2a4270d/pkg/kubelet/cm/devicemanager/manager.go#L325) and triggers when the device plugin sends updates to the kubelet (see https://github.com/kubernetes/kubernetes/blob/a2e12354edefc420e829973721030c5cb2a4270d/pkg/kubelet/cm/devicemanager/endpoint.go#L87 >>> https://github.com/kubernetes/kubernetes/blob/a2e12354edefc420e829973721030c5cb2a4270d/pkg/kubelet/cm/devicemanager/endpoint.go#L96)